AirCasting/CoreData/MeasurementStreamStorage.swift

```diff
@@ -41,18 +42,26 @@ final class CoreDataMeasurementStreamStorage: MeasurementStreamStorage {
         self.context = persistenceController.editContext()
     }
     
-    func save() throws {
+    private var isSafelyAccessed = false
+    
+    /// All actions performed on CoreDataMeasurementStreamStorage must be performed
+    /// within a block passed to this methood.
+    /// This ensures thread-safety by dispathing all calls to the queue owned by the NSManagedObjectContext.
+    func accessStorage(_ task: @escaping(MeasurementStreamStorage) -> Void) {
         context.perform {
-            do {
-                try self.context.save()
-            } catch {
-                Log.info("Couldn't save the context: \(self.context)")
-            }
+            self.isSafelyAccessed = true
```

Comment:
this is quite clever 🧠 But I don't see why can't we just do `context.perform { }` inside each method instead

Reply:
That's what I originally intended, but I eventually decided this is a better way.
Having `perform` in every method makes them all asynchronous and so we can no longer throw from them. Instead, I'd need to add `completionBlocks` to all of them. Other callers usually call a few methods of `MeasurementStreamStorage` in sequence and thus we'd end up having pyramids of doom everywhere.🤷🏻‍♀️